presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -21,88 +22,372 @@
+                .filter(entry -> remoteClusterInfos.get(entry).isHealthy())
+                .collect(Collectors.toList());
+
+        if (healthyClusterURIs.isEmpty()) {
+            log.info("No healthy clusters found");
+            return Optional.empty();
         }
+        log.info("Healthy clusters: %s", healthyClusterURIs);
 
-        return scheduler.getDestination(requestInfo.getUser());
+        config.getScheduler().setCandidates(healthyClusterURIs);
+        if (config.getSchedulerType() == WEIGHTED_RANDOM_CHOICE) {
+            config.getScheduler().setWeights(config.getServerWeights().get(groupSpec.getName()));
+        }
+        return config.getScheduler().getDestination(requestInfo.getUser());
     }
 
     private Optional<String> matchGroup(RequestInfo requestInfo)
     {
-        return groupSelectors.stream()
+        return currentConfig.get().groupSelectors.stream()
                 .map(s -> s.match(requestInfo))
                 .filter(Optional::isPresent)
                 .map(Optional::get)
                 .findFirst();
     }
 
-    private void initializeServerWeights()
+    private void initializeMembersDiscoveryURI(Map<URI, URI> discoveryURIs)
     {
-        groups.forEach((name, groupSpec) -> {
+        currentConfig.get().getGroups().forEach((name, groupSpec) -> {
             List<URI> members = groupSpec.getMembers();
-            List<Integer> weights = groupSpec.getWeights();
-            serverWeights.put(name, new HashMap<>());
+            List<URI> membersDiscoveryURI = groupSpec.getMembersDiscoveryURI();
             for (int i = 0; i < members.size(); i++) {
-                serverWeights.get(name).put(members.get(i), weights.get(i));
+                discoveryURIs.put(members.get(i), membersDiscoveryURI.get(i));
             }
         });
     }
+
+    @VisibleForTesting
+    public ConcurrentHashMap<URI, RemoteClusterInfo> getRemoteClusterInfos()
+    {
+        return remoteClusterInfos;
+    }
+
+    @VisibleForTesting
+    public ConcurrentHashMap<URI, RemoteQueryInfo> getRemoteQueryInfos()
+    {
+        return remoteQueryInfos;
+    }
+
+    @VisibleForTesting
+    public boolean getIsWatchServiceStarted()
+    {
+        return isWatchServiceStarted.get();
+    }
+
+    public interface OnConfigChangeDetection
+    {
+        void apply();
+    }
+
+    public static class ClusterStatusTracker
+    {
+        private final Logger log = Logger.get(com.facebook.presto.router.cluster.ClusterManager.ClusterStatusTracker.class);
+
+        private final ClusterManager clusterManager;
+        private final ScheduledExecutorService queryInfoUpdateExecutor;
+
+        @Inject
+        public ClusterStatusTracker(ClusterManager clusterManager)
+        {
+            this.clusterManager = requireNonNull(clusterManager, "clusterManager is null");
+            this.queryInfoUpdateExecutor = newSingleThreadScheduledExecutor(threadsNamed("query-info-poller-%s"));
+        }
+
+        @PostConstruct
```

Comment:
I've read [mixed opinions](https://github.com/google/guice/issues/62) about how Guice will handle `PostConstruct` annotated methods. Can you just remove this annotation and simply call `startPollingQueryInfo()` from the constructor instead ? This removes any ambiguity around lifecycle management